awx/main/access.py

```diff
@@ -233,16 +240,58 @@ def can_copy_related(self, obj):
         '''
         return True
 
-    def can_attach(self, obj, sub_obj, relationship, data,
-                   skip_sub_obj_read_check=False):
-        if skip_sub_obj_read_check:
-            return self.can_change(obj, None)
+    def get_attach_access(self, obj, sub_obj, relationship, skip_read_check=False, disassociate=False):
+        '''Directionality for attach permission is maintained here
+        Example:
+            Does user have permission to add A to "parents" of B?
+        An attach access class exists for model A "children" relationship
+        This will be automatically reversed to
+            Does user have permission to add B to "children" of A?
+        '''
+        if '.' in relationship:
+            hop, relationship = relationship.split('.', 1)
+            obj = getattr(obj, hop)
+        field = obj._meta.get_field(relationship)
+        if isinstance(field, ManyToOneRel):
+            # this call is in the wrong place, not an attachment but a modification
+            access = access_registry[sub_obj.__class__](self.user)
+            return access.can_change(sub_obj, {field.remote_field.name: obj})
+        if not isinstance(field, (ManyToManyField, ManyToManyRel)):
+            raise RuntimeError('Cannot call this with a {} relationship obj:{} rel:{}.'.format(
+                field.__class__.__name__, obj.__class__.__name__, relationship))
+        through_model = getattr(obj, relationship).through
+        if through_model not in attach_registry:
+            raise RuntimeError('Attchment logic for {}<-{} via {} has not be written'.format(
```

Comment:
Typo: 'Attachment'

Reply:
Though, should these `RuntimeError`s be `PermissionError`s instead?

Reply:
The idea is that these errors shouldn't happen, and forcing a 500 error makes it more likely that we'll see them, so we'll be alerted, so that we'll fix them. This message, in particular, is something that I definitely don't want to show to the user. The only alternative I think is acceptable is a generic PermissionDenied(), to simply error on the side of denying the user access. Historically, we've done that in some of these methods when the call arguments were not expected.

Reply:
I like this idea *IF* we have a good way to catch this via tests or some other interface.

Reply:
@matburt I have added a unit test that runs all views through this logic, assuring that exceptions are not raised for any of them.